src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs

```diff
@@ -171,5 +174,62 @@ public static SemanticMap GetSemanticMap(this SemanticModel semanticModel, Synta
 
             return new TokenSemanticInfo(declaredSymbol, aliasSymbol, allSymbols, type, convertedType, token.Span);
         }
+
+        public static string GenerateNameFromType(this SemanticModel semanticModel, ITypeSymbol type, ISyntaxFacts syntaxFacts, bool capitalize)
+        {
+            var pluralize = semanticModel.Pluralize(type);
+            var typeArguments = type.GetAllTypeArguments();
+
+            // We may be able to use the type's arguments to generate a name if we're working with an enumerable type.
+            if (pluralize && TryGenerateNameFromTypeArgument(syntaxFacts, typeArguments, capitalize, out var typeArgumentParameterName))
+            {
+                return typeArgumentParameterName;
+            }
+
+            // If there's no type argument and we have an array type, we should pluralize, e.g. frogs[] instead of frog[]
+            if (type.TypeKind == TypeKind.Array && typeArguments.IsEmpty)
+            {
+                return type.CreateParameterName(capitalize).Pluralize();
+            }
+
+            // Otherwise assume no pluralization, e.g. using 'immutableArray', 'list', etc. instead of their
+            // plural forms
+            return type.CreateParameterName(capitalize);
+        }
+
+        private static bool Pluralize(this SemanticModel semanticModel, ITypeSymbol type)
+        {
+            if (type == null)
+                return false;
+
+            if (type.SpecialType == SpecialType.System_String)
```

Comment:
this isappropriate.  however, it's worth doc'ing why we need to special case.

Reply:
I just moved the method over and I'm actually not quite sure why this string case is needed. Could you possibly elaborate?

Reply:
Ohhh. Is it because string implements System.Collections.Generic.IEnumerable<char>?

Reply:
bingo :)